### PR TITLE
[Windows] Allow to customize the Shell FlyoutIcon 

### LINKED ISSUE #7577: FlyoutIcon does not work on Windows, although it works on Android
### Description

I am taking a course about .NET MAUI in Microsoft Learn (https://docs.microsoft.com/en-us/learn/dotnet-maui/create-multi-page-apps/3-exercise-implement-flyout-navigation).... I did the exercise, where you have to specify an icon to replace the hamburger menu. When I run the example, it works fine on Android showing the new icon, but on Windows, it still shows the hamburger menu.

I uploaded 2 images: one for Android and another for Windows => https://postimg.cc/gallery/p2JShgP

### Steps to Reproduce

Follow 3rd step of this page: [Microsoft Learn MAUI Course](https://docs.microsoft.com/en-us/learn/dotnet-maui/create-multi-page-apps/3-exercise-implement-flyout-navigation)

Add a FlyoutIcon property to the <Shell> node to display an image. By default it will display 3 horizontal bars, but we can change it to be whatever we like. The markup should look like this:

```xaml
<Shell
    x:Class="Astronomy.AppShell"
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    xmlns:local="clr-namespace:Astronomy.Pages"
    FlyoutIcon="moon.png">
```

### Version with bug

6.0 (current)

### Last version that worked well

Release Candidate 3 (current)

### Affected platforms

Windows

### Affected platform versions

Windows 11

### Did you find any workaround?

I dont know a workaround for this error.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Allow to customize the Shell FlyoutIcon on Windows.

image

### Issues Fixed

Fixes #7577

### CODE REVIEW

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs [hunk 1]
```diff
@@ -124,6 +126,21 @@ public static void MapFlyoutBackground(ShellHandler handler, Shell view)
...
+			var togglePaneButton = handler.PlatformView.TogglePaneButton;
+
+			if (togglePaneButton is null)
+				return;
+
+			var provider = handler.GetRequiredService<IImageSourceServiceProvider>();
+			var flyoutIcon = view.FlyoutIcon;
+
+			togglePaneButton
+				.UpdateBackgroundImageAsync(flyoutIcon, provider)
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs]: Applied some changes.
fix-shell-icon

"I also noticed that after this PR I can no longer click on the toggle button.". What do you mean?, cannot tap the toggle button to open/close the menu?

[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs]: @PureWeen Could you review it again?. Thanks in advance.

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs [hunk 2]
```diff
@@ -270,5 +272,22 @@ public static void UpdateFlyoutWidth(this MauiNavigationView navigationView, IFl
 			// At some point this Template Setting is going to show up with a bump to winui
 			//handler.PlatformView.OpenPaneLength = handler.PlatformView.TemplateSettings.OpenPaneWidth;
 		}
+
+		public static async Task UpdateFlyoutIconAsync(this AnimatedIcon platformView, IImageSource? imageSource, IImageSourceServiceProvider? provider)
+		{
+			if (platformView is null)
+				return;			
+
+			if (provider is not null && imageSource is not null)
```
Comment (PureWeen) [inline src/Core/src/Platform/Windows/NavigationViewExtensions.cs:281]: If the user hasn't set the `icon` can we return it back to the font glyph?

https://github.com/microsoft/microsoft-ui-xaml/blob/a7183df20367bc0e2b8c825430597a5c1e6871b6/dev/NavigationView/NavigationView_rs1_themeresources.xaml#L389-L391

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs [hunk 3]
```diff
@@ -270,5 +272,22 @@ public static void UpdateFlyoutWidth(this MauiNavigationView navigationView, IFl
...
+			{
+				var service = provider.GetRequiredImageSourceService(imageSource);
+				var nativeImageSource = await service.GetImageSourceAsync(imageSource);
+
+				platformView.Source = null;
```
Comment (PureWeen) [inline src/Core/src/Platform/Windows/NavigationViewExtensions.cs:286]: What are your thoughts on changing the height/width of the AnimatedIcon? Right now, it forces the width and height to 16 

https://github.com/microsoft/microsoft-ui-xaml/blob/a7183df20367bc0e2b8c825430597a5c1e6871b6/dev/NavigationView/NavigationView_rs1_themeresources.xaml#L380-L381

So, the Icon is super tiny. 

I'm wondering if we should just set the width and height of the `AnimatedIcon` to `Double.Nan`

This sizes the book icon to a bit better of a size.
image

It looks like the outer container constrains the icon so even if the user loads in a massive image it'll still shrink.

Here's the .NET MAUI Bot inside the sandbox

image

File: src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs [hunk 4]
```diff
@@ -124,6 +127,20 @@ public static void MapFlyoutBackground(ShellHandler handler, Shell view)
...
+			var togglePaneButton = handler.PlatformView.TogglePaneButton;
+
+			if (togglePaneButton is null)
+				return;
+
+			var icon = togglePaneButton.GetFirstDescendant<AnimatedIcon>();
+			var flyoutIcon = view.FlyoutIcon;
+			var provider = handler.GetRequiredService<IImageSourceServiceProvider>();
+
+			icon.UpdateFlyoutIconAsync(flyoutIcon, provider).FireAndForget();
```
Comment (PureWeen) [inline src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs:141]: Can we make the extension method against `MauiNavigationView` and then `UpdateFlyoutIcon` will just be against `MauiNavigationView` ? I think that'll make it easier to reuse if folks want to reuse the extension method.

**Review by PureWeen — CHANGES_REQUESTED** [review]
Can we add some tests as well?

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs (hunk 3 above)
Comment (jsuarezruiz) [reply src/Core/src/Platform/Windows/NavigationViewExtensions.cs:286]: The container limits the size, I think that's fine.

File: src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs (hunk 4 above)
Comment (jsuarezruiz) [reply src/Controls/src/Core/Handlers/Shell/ShellHandler.Windows.cs]: Yeah, would be easier to extend. Added.

File: src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.Windows.cs [hunk 5]
```diff
@@ -27,6 +28,43 @@ protected Task CheckFlyoutState(ShellHandler handler, bool desiredState)
...
+			{
+				shell.FlyoutBehavior = FlyoutBehavior.Flyout;
+				shell.FlyoutIcon = "red.png";
+
+				var shellItem = new FlyoutItem();
+				shellItem.Items.Add(new ContentPage());
+				shell.Items.Add(shellItem);
+			});
+
+			await InvokeOnMainThreadAsync(async () =>
```
Comment (PureWeen) [inline src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.Windows.cs:46]: `CreateHandleAndAddToWindow` already invokes on the main thread so you can remove this part.

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs [hunk 6]
```diff
@@ -270,5 +272,47 @@ public static void UpdateFlyoutWidth(this MauiNavigationView navigationView, IFl
...
+		public static async Task UpdateFlyoutIconAsync(this MauiNavigationView navigationView, IImageSource? imageSource, IImageSourceServiceProvider? provider)
+		{
+			var togglePaneButton = navigationView.TogglePaneButton;
+
+			if (togglePaneButton is null)
+				return;
+
+			var animatedIcon = togglePaneButton.GetFirstDescendant<AnimatedIcon>();
+			
+			if (animatedIcon is null)
```
Comment (PureWeen) [inline src/Core/src/Platform/Windows/NavigationViewExtensions.cs:285]: I'm running this locally and this is always null for me, and the icon isn't changing.

I tested on the sandbox

```XAML
<Shell
    xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
    x:Class="Maui.Controls.Sample.SandboxShell"
    xmlns:local="clr-namespace:Maui.Controls.Sample"
    x:Name="shell"
    FlyoutIcon="dotnet_bot.png">
```

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.Windows.cs (hunk 5 above)
Comment (jsuarezruiz) [reply src/Controls/tests/DeviceTests/Elements/Shell/ShellTests.Windows.cs]: Right, removed.

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs (hunk 6 above)
Comment (jsuarezruiz) [reply src/Core/src/Platform/Windows/NavigationViewExtensions.cs:285]: Have done some tests and I think have been an error doing fallback to the original icon. Fixed here https://github.com/dotnet/maui/pull/14663/commits/7766d4c9c356ac6542336b27566ef049bb9ef85f

Tested on the Gallery and with Device tests.

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs [hunk 7]
```diff
@@ -270,5 +272,48 @@ public static void UpdateFlyoutWidth(this MauiNavigationView navigationView, IFl
...
+				var nativeImageSource = await service.GetImageSourceAsync(imageSource);
+
+				platformView.Source = null;
+				
+				var fallbackIconSource = new ImageIconSource { ImageSource = nativeImageSource?.Value };
+
+				platformView.Height = platformView.Width = double.NaN;
+				platformView.FallbackIconSource = fallbackIconSource;
+			}
+			else
```
Comment (PureWeen) [inline src/Core/src/Platform/Windows/NavigationViewExtensions.cs:309]: I think we still need to restore the `Source` on the platformView. When I tested this with hot reload working. The ICON was blank when returning back to the default.

Something like this work?

```C#
else if (platformView.Source is null)
{
	// Fallback to the default hamburger icon
	// https://github.com/microsoft/microsoft-ui-xaml/blob/a7183df20367bc0e2b8c825430597a5c1e6871b6/dev/NavigationView/NavigationView_rs1_themeresources.xaml#L389-L391
	var fallbackIconSource = new FontIconSource { Glyph = "&#xE700;" };
	platformView.Height = platformView.Width = 16d;
	platformView.FallbackIconSource = fallbackIconSource;
	platformView.Source = new AnimatedGlobalNavigationButtonVisualSource();
}
```

**Review by PureWeen — CHANGES_REQUESTED** [review]

File: src/Core/src/Platform/Windows/NavigationViewExtensions.cs (hunk 7 above)
Comment (PureWeen) [reply src/Core/src/Platform/Windows/NavigationViewExtensions.cs:309]: My other thought would be to just store this Animated Icon on the `MauiNavigationView` itself. 

Then you can just swap back to it?

Comment (jsuarezruiz) [reply src/Core/src/Platform/Windows/NavigationViewExtensions.cs:310]: > I think we still need to restore the `Source` on the platformView. When I tested this with hot reload working. The ICON was blank when returning back to the default.
> 
> Something like this work?
> 
> ```cs
> else if (platformView.Source is null)
> {
> 	// Fallback to the default hamburger icon
> 	// https://github.com/microsoft/microsoft-ui-xaml/blob/a7183df20367bc0e2b8c825430597a5c1e6871b6/dev/NavigationView/NavigationView_rs1_themeresources.xaml#L389-L391
> 	var fallbackIconSource = new FontIconSource { Glyph = "&#xE700;" };
> 	platformView.Height = platformView.Width = 16d;
> 	platformView.FallbackIconSource = fallbackIconSource;
> 	platformView.Source = new AnimatedGlobalNavigationButtonVisualSource();
> }
> ```

Right, updated and tested.

**Review by PureWeen — APPROVED** [review]

**Pull request merged**